framework/src/mesh/AnnularMesh.C

```diff
@@ -0,0 +1,230 @@
+    _rmax(getParam<Real>("rmax")),
+    _tmin(getParam<Real>("tmin")),
+    _tmax(getParam<Real>("tmax")),
+    _growth_r(getParam<Real>("growth_r")),
+    _len(_growth_r == 1.0 ? (_rmax - _rmin) / _nr
+                          : (_rmax - _rmin) * (1.0 - _growth_r) / (1.0 - std::pow(_growth_r, _nr))),
+    _full_annulus(_tmin == 0.0 && _tmax == 2 * M_PI)
+{
+  // catch likely user errors
+  if (_rmax <= _rmin)
+    mooseError("AnnularMesh: rmax must be greater than rmin");
+  if (_tmax <= _tmin)
+    mooseError("AnnularMesh: tmax must be greater than tmin");
+}
+
+Real
+AnnularMesh::getMinInDimension(unsigned int component) const
+{
+  switch (component)
+  {
+    case 0:
+      return -_rmax;
+    case 1:
+      return -_rmax;
+    case 2:
+      return 0.0;
+    default:
+      mooseError("Invalid component");
+  }
+}
+
+Real
+AnnularMesh::getMaxInDimension(unsigned int component) const
+{
+  switch (component)
+  {
+    case 0:
+      return _rmax;
+    case 1:
+      return _rmax;
+    case 2:
+      return 0.0;
+    default:
+      mooseError("Invalid component");
+  }
+}
+
+MooseMesh &
+AnnularMesh::clone() const
+{
+  return *(new AnnularMesh(*this));
+}
+
+void
+AnnularMesh::buildMesh()
+{
+  const Real dt = (_tmax - _tmin) / _nt;
+
+  MeshBase & mesh = getMesh();
+  mesh.clear();
+  mesh.set_mesh_dimension(2);
+  mesh.set_spatial_dimension(2);
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+
+  const unsigned num_angular_nodes = (_full_annulus ? _nt : _nt + 1);
+  const unsigned num_nodes =
+      (_rmin > 0.0 ? (_nr + 1) * num_angular_nodes : _nr * num_angular_nodes + 1);
+  const unsigned min_nonzero_layer_num = (_rmin > 0.0 ? 0 : 1);
+  std::vector<Node *> nodes(num_nodes);
+  unsigned node_id = 0;
+
+  // add nodes at rmax that aren't yet connected to any elements
+  Real current_r = _rmax;
+  for (unsigned angle_num = 0; angle_num < num_angular_nodes; ++angle_num)
+  {
+    const Real angle = _tmin + angle_num * dt;
+    const Real x = current_r * std::cos(angle);
+    const Real y = current_r * std::sin(angle);
+    nodes[node_id] = mesh.add_point(Point(x, y, 0.0), node_id);
+    node_id++;
+  }
+
+  // add nodes at smaller radii, and connect them with elements
+  for (unsigned layer_num = _nr; layer_num > min_nonzero_layer_num; --layer_num)
+  {
+    if (layer_num == 1)
+      current_r = _rmin; // account for precision loss
+    else
+      current_r -= _len * std::pow(_growth_r, layer_num - 1);
+
+    // add node at angle = _tmin
+    nodes[node_id] = mesh.add_point(
+        Point(current_r * std::cos(_tmin), current_r * std::sin(_tmin), 0.0), node_id);
+    node_id++;
+    for (unsigned angle_num = 1; angle_num < num_angular_nodes; ++angle_num)
+    {
+      const Real angle = _tmin + angle_num * dt;
+      const Real x = current_r * std::cos(angle);
+      const Real y = current_r * std::sin(angle);
+      nodes[node_id] = mesh.add_point(Point(x, y, 0.0), node_id);
+      Elem * elem = mesh.add_elem(new Quad4);
```

Comment:
I don't see where you are adding this element to a subdomain. I guess the default is zero. It might be useful to allow users to assign their own subdomains but it's not critical.

Reply:
It's easy to let users define the subdomain numbers, and i'll do that.  GeneratedMesh doesn't allow this, and i was just following its lead.